### PR TITLE
Add ViViT implementation

### PR DESCRIPTION
**Pull request by abhi-glitchhg**
Here is a model3 of ViViT

cc @NeelayS @aditya-agrawal-30502

### CODE REVIEW

File: vformer/encoder/vivit.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        b, n, s, d = x.shape
+        x = torch.flatten(x, start_dim=0, end_dim=1)  # 1×nt·nh·nw·d --> nt×nh·nw·d
+
+        x = self.spatial_attention(x) + x
+
+        x = x.reshape(b, n, s, d).transpose(1, 2)
+        x = torch.flatten(x, start_dim=0, end_dim=1)  # nt×nh·nw·d --> nh·nw×nt·d
+
+        x = self.temporal_attention(x) + x
+
```
Comment (abhi-glitchhg) [inline vformer/encoder/vivit.py:37]: I forgot to add layer norm in these; will update this

**Review by abhi-glitchhg — COMMENTED** [review]
.

File: tests/test_encoder.py [hunk 2]
```diff
@@ -162,3 +162,15 @@ def test_TubeletEmbedding():
     out = embedding(test_tensor)
     assert out.shape == (11, 3, 16, 96)
     del embedding
+
+
+def test_ViViTEncoder():
+    encoder = ENCODER_REGISTRY.get("ViViTEncoder")(
```
Comment (NeelayS) [inline tests/test_encoder.py:168]:
```suggestion

    encoder = ENCODER_REGISTRY.get("ViViTEncoder")(
```

File: tests/test_encoder.py [hunk 3]
```diff
@@ -162,3 +162,15 @@ def test_TubeletEmbedding():
...
+        dim=192, num_heads=3, head_dim=64, p_dropout=0.0, depth=3
+    )
+
+    test_tensor = torch.randn(7, 20, 196, 192)
+
```
Comment (NeelayS) [inline tests/test_encoder.py:173]:
```suggestion
```

File: tests/test_encoder.py [hunk 4]
```diff
@@ -162,3 +162,15 @@ def test_TubeletEmbedding():
...
+    logits = encoder(test_tensor)
+
```
Comment (NeelayS) [inline tests/test_encoder.py:175]:
```suggestion
```

File: tests/test_models.py [hunk 5]
```diff
@@ -511,11 +511,52 @@ def test_ViViT():
...
+        embedding_dim=512,
+        depth=3,
+        num_heads=4,
+        head_dim=32,
+        p_dropout=0.0,
+        in_channels=3,
+    )
+    test_tensor3 = torch.randn(32, 32, 3, 64, 64)
+    logits = model(test_tensor3)
+
```
Comment (NeelayS) [inline tests/test_models.py:540]:
```suggestion
```

File: tests/test_models.py [hunk 6]
```diff
@@ -511,11 +511,52 @@ def test_ViViT():
...
+    assert logits.shape == (32, 10)
+
```
Comment (NeelayS) [inline tests/test_models.py:542]:
```suggestion
```

File: vformer/models/classification/vivit.py [hunk 7]
```diff
@@ -109,3 +145,104 @@ def forward(self, x):
...
+                embedding_dim,
+            ],
+            n_classes=num_classes,
+        )
+
+    def forward(self, x):
+
+        x = self.tubelet_embedding(x)
+        x = self.pos_embbedding(x)
+
```
Comment (NeelayS) [inline vformer/models/classification/vivit.py:242]:
```suggestion
```

File: vformer/models/classification/vivit.py [hunk 8]
```diff
@@ -109,3 +145,104 @@ def forward(self, x):
...
+        x = self.encoder(x)
+
```
Comment (NeelayS) [inline vformer/models/classification/vivit.py:244]:
```suggestion
```

File: vformer/models/classification/vivit.py [hunk 9]
```diff
@@ -109,3 +145,104 @@ def forward(self, x):
...
+        x = x.mean(dim=1)
+
```
Comment (NeelayS) [inline vformer/models/classification/vivit.py:246]:
```suggestion
```

File: vformer/models/classification/vivit.py [hunk 10]
```diff
@@ -109,3 +145,104 @@ def forward(self, x):
...
+        x = self.decoder(x)
+        return x
```
Comment (NeelayS) [inline vformer/models/classification/vivit.py:248]:
```suggestion

        return x
```

File: vformer/encoder/vivit.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                    dim, num_heads, head_dim, p_dropout, out_dim, hidden_dim
+                )
+            )
+
+    def forward(self, x):
+        b = x.shape[0]
+        for blk in self.encoder:
+            x = blk(x)
+        x = x.reshape(b, -1, x.shape[-1])
+        return x
```
Comment (NeelayS) [inline vformer/encoder/vivit.py:70]:
```suggestion

        return x
```

File: vformer/encoder/vivit.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+        for _ in range(depth):
+            self.encoder.append(
+                ViViTEncoderBlock(
+                    dim, num_heads, head_dim, p_dropout, out_dim, hidden_dim
+                )
+            )
+
+    def forward(self, x):
+        b = x.shape[0]
```
Comment (NeelayS) [inline vformer/encoder/vivit.py:66]:
```suggestion

        b = x.shape[0]
```

File: vformer/encoder/vivit.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        for blk in self.encoder:
```
Comment (NeelayS) [inline vformer/encoder/vivit.py:67]:
```suggestion

        for blk in self.encoder:
```

File: vformer/encoder/vivit.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+            x = blk(x)
+        x = x.reshape(b, -1, x.shape[-1])
```
Comment (NeelayS) [inline vformer/encoder/vivit.py:69]:
```suggestion

        x = x.reshape(b, -1, x.shape[-1])
```

File: vformer/encoder/vivit.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            dim=dim, fn=VanillaSelfAttention(dim, num_heads, head_dim, p_dropout)
+        )
+        self.spatial_attention = PreNorm(
+            dim=dim, fn=VanillaSelfAttention(dim, num_heads, head_dim, p_dropout)
+        )
+
+        self.mlp = FeedForward(dim=dim, hidden_dim=hidden_dim, out_dim=out_dim)
+
+    def forward(self, x):
+        b, n, s, d = x.shape
```
Comment (NeelayS) [inline vformer/encoder/vivit.py:29]:
```suggestion

        b, n, s, d = x.shape
```

**Review by NeelayS — CHANGES_REQUESTED** [review]

**Pull request merged**